### ISSUE TITLE
Update mock component to allow setting values on a parent reactive form via the component

### ISSUE THREAD

**Issue by vespertilian** [ike18t/ng-mocks#11]
Hi, Nice Library!

Would you be open to updating your mock component to allow Reactive Form values to be propagated out. I like to drive the reactive form, via the form element component, when I am testing so that I know that it is bound correctly. 

Something like this:

#### Sample Class
```ts
export class MockReactiveFormsElement implements ControlValueAccessor {
  propagateChange = (_: any) => {};
  setTouched = (_: any) => {};

  writeValue(value: any) { }

  registerOnChange(fn) {
    this.propagateChange = fn;
  }

  registerOnTouched(fn) {
    this.setTouched = fn;
  }

  setDisabledState(isDisabled: boolean) {}
}
```

#### Use in your test like this:

```ts
const typeSelect: MockMdSelectComponent = fixture.debugElement.query(By.css('md-select')).componentInstance;
typeSelect.propagateChange('Ninja');
expect(component.warriorForm.value.type).toEqual('Ninja');
```

Maybe use `__propagatesChange` or some type prefix so as not to clash with a component that has that method, or it could be a subclass that has a seperate function `mockReactiveFormComponent`

Happy to submit a pull request if you will merge. Just let me know your preferred approach.

**Comment by getsaf** [comment]
Thanks for the request, I have a *little* experience with Reactive Forms but I'm not following your question. Do you mind setting up a quick StackBlitz project to describe the issue? Don't worry about creating test in the StackBlitz, just a component that is representative of this issue so I can follow along with this feature request.

Thanks!

**Comment by vespertilian** [comment]
Here it is

https://stackblitz.com/github/vespertilian/testing-reactive-forms-third-party-components

Might be easier to pull it and run it from the CLI 

https://github.com/vespertilian/testing-reactive-forms-third-party-components

```ts
  it('should update the form when you fill in the type "select" field', () => {
    const {fixture, component} = setup();
    const typeSelect: MockMdSelectComponent = fixture.debugElement.query(By.css('mat-select')).componentInstance;

    typeSelect.propagateChange('Ninja');
    expect(component.warriorForm.value.type).toEqual('Ninja');
  });
```

Basically when testing forms I prefer to drive the form not just use for formController instance in the component controller as that does not tell you if you have bound to the form via "formControlName"

**Comment by vespertilian** [comment]
@getsaf Just checking in, you are probably busy, which is fine. If there is anything else I can do to help move this issue along please let me know.

**Comment by ike18t** [comment]
@vespertilian What do you think about writeValue triggering the change callback? Would that solve your problem?

**Comment by vespertilian** [comment]
Hey @ike18t, sorry for the late reply this slipped by me. 

I had thought that as well initially, it's just as write value is normally used to set the value on the custom form control, I was concerned this will lead to unexpected behaviour, or hard to discover bugs. 

I think this would work when you just grab the custom form control from the debug element and call writeValue.

However say you, or some other part of your test calls myFormControl.setValue(name: 'vespert'), if we bound it to writeValue on the form control, the value will be propagated back out, which would not usually happen as write value is normally only called if something inside the custom form control changes the value, so if you were say subscribing to value changes you would get a false output. 

Also not sure if binding the input and output together could lead to some type of infinite feedback loop / weird change detection issues. 

```
interface ControlValueAccessor {
  writeValue(obj: any): void
  registerOnChange(fn: any): void
  registerOnTouched(fn: any): void
  setDisabledState(isDisabled: boolean)?: void
}

# it's these two 'fn' we want to be able to mock out and manually activate 
registerOnChange(fn: any): void
registerOnTouched(fn: any): void
```

We could do it with a getter and setter on value and touched maybe? 

```
formControl.value = 'vespert'
```

I still think 

```
  propagateChange = (_: any) => {};
  setTouched = (_: any) => {};
  // or
  __propagateChange = (_: any) => {};
  __setTouched = (_: any) => {};
```

is the clearest and cleanest way to accomplish this, I would probably go with the __ just in case someone has a function on there custom form control called propagateChange.

**Comment by ike18t** [comment]
@vespertilian Fair enough. I don't have that much experience dealing with reactive forms so I'm going to have to lean on your experience.  Feel free to submit a PR.

**Comment by vespertilian** [comment]
@ike18t Will do! Hopefully I can find some time this weekend.

**Comment by vespertilian** [comment]
@ike18t No internet this weekend, it will have to be some time next week!

**Comment by ike18t** [comment]
@vespertilian A weekend w/o internet?  Are you OK?  Did you survive?

**Comment by vespertilian** [comment]
Yeah I had a great weekend, I was actually at a JavaScript camp. https://ix.campjs.com. So I figured it would be a good place to get the work done but every-time I tried to go to Github the 100+ other devs were also trying to download something, so I just gave up on the third timeout attempt. 

I could have tried harder, but there were also waterfall walks and mountain vistas to checkout. 

https://twitter.com/campjs

![img_1037](https://user-images.githubusercontent.com/441418/44123401-80342cba-a06b-11e8-9b4c-856c88c8c667.jpg)
